shopify_function_macro/src/lib.rs

```diff
@@ -114,6 +114,150 @@ pub fn shopify_function(
     gen.into()
 }
 
+#[derive(Clone, Default)]
+struct ShopifyFunctionTargetArgs {
+    query_path: Option<LitStr>,
+    schema_path: Option<LitStr>,
+    input_stream: Option<Expr>,
+    output_stream: Option<Expr>,
+}
+
+impl ShopifyFunctionTargetArgs {
+    fn parse<K: syn::parse::Parse, V: syn::parse::Parse>(
+        input: &ParseStream<'_>,
+    ) -> syn::Result<V> {
+        let _ = input.parse::<K>()?;
+        let _ = input.parse::<Token![=]>()?;
+        let value: V = input.parse()?;
+        if input.lookahead1().peek(Token![,]) {
+            input.parse::<Token![,]>()?;
+        }
+        Ok(value)
+    }
+}
+
+impl Parse for ShopifyFunctionTargetArgs {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let mut args = Self::default();
+        while !input.is_empty() {
+            let lookahead = input.lookahead1();
+            if lookahead.peek(kw::query_path) {
+                args.query_path = Some(Self::parse::<kw::query_path, LitStr>(&input)?);
+            } else if lookahead.peek(kw::schema_path) {
+                args.schema_path = Some(Self::parse::<kw::schema_path, LitStr>(&input)?);
+            } else if lookahead.peek(kw::input_stream) {
+                args.input_stream = Some(Self::parse::<kw::input_stream, Expr>(&input)?);
+            } else if lookahead.peek(kw::output_stream) {
+                args.output_stream = Some(Self::parse::<kw::output_stream, Expr>(&input)?);
+            } else {
+                panic!("Unknown shopify_function_target parameter");
+            }
+        }
+        Ok(args)
+    }
+}
+
+fn extract_shopify_function_return_type(ast: &syn::ItemFn) -> &syn::Ident {
+    use syn::*;
+
+    let ReturnType::Type(_arrow, ty) = &ast.sig.output else {
+        panic!("Shopify Function require an explicit return type")
```

Comment:
I wanted to do that, but was surprised to find that I was struggling to create my own `syn::Error` which I need to do to use `syn::Result`. Do you know of a way?

Reply:
[`syn::Error::new`](https://docs.rs/syn/2.0.28/syn/struct.Error.html#method.new) with the span from calling [`syn::spanned::Spanned::span`](https://docs.rs/syn/2.0.28/syn/spanned/trait.Spanned.html) to get a span (or [`proc_macro2::Span::call_site`](https://docs.rs/proc-macro2/1.0.66/proc_macro2/struct.Span.html#method.call_site) if there is nothing that implements `Spanned` and you want to use the macro invocation span instead. Or you could use [`syn::Error::new_spanned`](https://docs.rs/syn/2.0.28/syn/struct.Error.html#method.new_spanned) on something `quote::ToTokens` to avoid calling `span`.

[Here](https://github.com/adampetro/bluejay/blob/c6abb5390a0d0f16781fcb7cba50497e5d6b505d/bluejay-typegen-macro/src/lib.rs#L111-L148) are a few examples from a project of mine (coincidentally a GraphQL typegen macro crate 😛).